mm2src/coins/utxo.rs

```diff
@@ -1555,6 +1563,8 @@ fn spawn_server_version_retry_loop(weak_client: Weak<ElectrumClientImpl>, client
 
             match client.set_protocol_version(&electrum_addr, actual_version).await {
                 Ok(()) => {
+                    // Move this server to the end of the connections list if it's the first since it disconnects and reconnects (unstable)
```

Comment:
Please correct me, but if I remember correctly, if we get to this condition, then we received, checked and set the protocol version successfully. I think then we can leave this electrum as the first because it's just the correct one.

Reply:
I have the same reflection, it's a bit unclear for me

Reply:
The idea was that if an electrum server disconnects and reconnects a lot https://github.com/KomodoPlatform/atomicDEX-API/blob/d15d043d48492715caf754e4e7ba8b731f64c5fa/mm2src/coins/utxo.rs#L1589-L1590 then it should be moved from first to last to avoid waiting for at least the first request of a multi request to timeout on every new multi request. https://github.com/KomodoPlatform/atomicDEX-API/blob/d15d043d48492715caf754e4e7ba8b731f64c5fa/mm2src/coins/utxo/rpc_clients.rs#L1245 I see that I missed the case where `negotiate_version` is false so server rotation should probably be done if a transport error is returned as @sergeyboyko0791 suggested above. Another solution can be to use `on_connect_rx` if `negotiate_version` is false to detect reconnections (Unstable connections). I will investigate both solutions and see which one should work better.

Reply:
From testing, it seems that electrum servers disconnect and reconnect a lot, this doesn't necessarily cause requests to fail and I think it would be excessive to rotate electrums in this case. As a result of that, I decided to move the failing electrums to the end of the list (rotate) only when a request to an electrum fails as a part of a multi request.

Reply:
> From testing, it seems that electrum servers disconnect and reconnect a lot

That is strange, where did you see this? Electrum server should disconnect on session timeout https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html#envvar-SESSION_TIMEOUT after 600s (default) if no new `server.ping` was sent during that time.
The only other reason for a disconnect are the resource limits: https://electrumx-spesmilo.readthedocs.io/en/latest/environment.html#resource-usage-limits ... if the dude that runs the server hasn't disabled them or configured them properly, the server will start to slow down the requests till it disconnects you completely ... but then, you are not able to connect back anyway.

Reply:
Yes, that seems strange. @shamardy Could you please provide more details on these reconnections?

Reply:
After more testing, @cipig is correct. I found that these reconnections happened only in unit tests because of https://github.com/KomodoPlatform/atomicDEX-API/blob/f8039b46bb4c93747bb33683104b7bb4e32e1856/mm2src/coins/utxo/utxo_tests.rs#L37 I was testing what would happen when the first electrum times out with an infinite loop in unit tests while closing and opening the connection to the electrum using TCPView and because the test ran long enough other electrums would disconnect and reconnect due to session timeout.

Reply:
For the electrum servers rotation, I still think we should rotate the servers that fail as part of a multi-request only. What 
I mean is we shouldn't move servers that fail a direct request https://github.com/KomodoPlatform/atomicDEX-API/blob/f8039b46bb4c93747bb33683104b7bb4e32e1856/mm2src/coins/utxo/rpc_clients.rs#L1277 to the end of the electrums list or that fail the `server.ping` requests https://github.com/KomodoPlatform/atomicDEX-API/blob/f8039b46bb4c93747bb33683104b7bb4e32e1856/mm2src/coins/utxo/rpc_clients.rs#L1266-L1272 as this will overcomplicate things especially when `spawn_ping` is false as this will cause the servers to rotate with every reconnection.